File: project/src/terminal_server/Events.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Threading;
using System.Text;
using System.IO;
using Game.Utils;
using System.Text.RegularExpressions;
using Newtonsoft.Json;

namespace Terminal
{
    public class Event
    {
        public string type = "";
        public string data = "";

        public Event()
        {
            type = "";
            data = "";
        }
        public Event(string t, string d)
        {
            type = t;
            data = d;
        }

        public List<Event> processEventAndResponse()
        {
            List<Event> events = new List<Event>();
            switch (type)
            {
                case "force_console_content_fetch":
                    events.Add(MakeConsoleContentChangeEvent());
                    break;
                case "run_code":
                    RunCode(data);
                    break;
                default: break;
            }
            return events;
        }

        public static Event MakeConsoleContentChangeEvent()
        {
            return new Event("console_content_change", Game.UI.MacroExploiterWindow.Instance.ConsoleContent);
        }
        public static void RunCode(string code)
        {
            Game.UI.MacroExploiterWindow.Instance.ExecuteCode(code);
        }
    }

    public class EventsPackage
    {
        public List<Event> events = new List<Event>();

        public EventsPackage()
        {
            
        }

        public void ClearEvents()
        {
            events.Clear();
        }

        public List<Event> ProcessEventsAndReply()
        {
            List<Event> revents = new List<Event>();
            foreach (Event e in events)
            {
                if(e != null)
                {
                    var newEvents = e.processEventAndResponse();
                    if (newEvents!=null) revents.AddRange(newEvents);
                }
            }
            return revents;
        }
        public void ProcessEventsAndGatherInSelf(EventsPackage pck)
        {
            List<Event> revents = pck.ProcessEventsAndReply();
            events.AddRange(revents);
        }

        public static EventsPackage FromJSON(string json)
        {
            return JsonConvert.DeserializeObject<EventsPackage>(json);
        }
        public string ToJSON()
        {
            return JsonConvert.SerializeObject(this, Formatting.None);
        }
    }
    public class GameEventsProvider
    {
        public EventsPackage eventsPackage;

        public GameEventsProvider()
        {
            eventsPackage = new EventsPackage();
        }
        
        public void ProcessEventsAndGatherInSelf(EventsPackage pck)
        {
            pck.ProcessEventsAndGatherInSelf(pck);
        }

        public void Initialize()
        {
            Game.UI.MacroExploiterWindow.Instance.OnConsoleContentChange += ConsoleContentChange;
        }

        public void ConsoleContentChange(string content)
        {
            eventsPackage.events.Add(Event.MakeConsoleContentChangeEvent());
        }
    }
}

File: x-scripts-editor/src/api/events.js
import { Exportable, FC } from "extripo";

export class Event extends Exportable{
    constructor(){
        super()
        this.type = ""
        this.data = ""
    }
    static new(){
        return new this()
    }
}

export class EventForceConsoleContentFetch extends Event{
    constructor(){
        super()
        this.type = "force_console_content_fetch"
    }
}
export class EventRunCode extends Event{
    constructor(code=""){
        super()
        this.type = "run_code"
        this.data = code
    }
}

export class EventsPackage extends Exportable{
    constructor(){
        super()
        this.configFields({
            events: FC.arrayOf(Event),
        })
        this.events = []
    }
    clear(){
        this.events.splice(0, this.events.length)
    }
    addEvent(e){
        this.events.push(e)
    }
}

export class TerminalEventsProvider{
    constructor(){
        this.eventsPackage = new EventsPackage()
    }
    forceConsoleContentFetch(){
        this.eventsPackage.addEvent(EventForceConsoleContentFetch.new())
    }
    runCode(code){
        this.eventsPackage.addEvent(new EventRunCode(code))
    }
}